### PR TITLE
CE-1665 Set up Flags extension

### PR DESCRIPTION
**Pull request by adamkarminski**
This is the first commit for the new Flags extension (hooray!). Remember, this is still **work in progress**, so don't expect all perfect solutions. For example - there is a code duplication in _flagAction_ and _flagTypeAction_ but this is intentional since we don't know yet if they'll diverge in the future.

**What does it include?**
- The extension setup
- Basic interface for adding and removing Flags and types of Flags
- A method to get all flags for a page **with parameters** (hooray!)
- Documentation of the controller

**What does it not include?**
- FULL DOCUMENTATION ( @lukaszkonieczny ! )
- Any front-end, so no flags rendering and no management GUI
- Records updating
- Proper error handling
- Caching of the GET responses

Have fun with Flags!

@Wikia/community-engineering

For schema update: @drozdo

### CODE REVIEW

File: extensions/wikia/Flags/Flags.setup.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+/**
+ * Wikia Flags Extension
+ * @author Adam Karmiński
+ *
+ */
+
+$dir = dirname(__FILE__) . '/';
```
Comment (Grunny) [inline extensions/wikia/Flags/Flags.setup.php]: Use `__DIR__` rather than `dirname(__FILE__)`.

File: extensions/wikia/Flags/Flags.setup.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Models
+ */
+$wgAutoloadClasses['Flags\Models\FlagsModel'] = $dir . 'models/FlagsModel.class.php';
+$wgAutoloadClasses['Flags\Models\Flag'] = $dir . 'models/Flag.class.php';
+$wgAutoloadClasses['Flags\Models\FlagParameter'] = $dir . 'models/FlagParameter.class.php';
+$wgAutoloadClasses['Flags\Models\FlagType'] = $dir . 'models/FlagType.class.php';
+
+$wgExtensionCredits['other'][] = [
```
Comment (Grunny) [inline extensions/wikia/Flags/Flags.setup.php]: As a general unwritten convention, the credits should go at the start of the setup file as a kind of documentation.

Comment (adamkarminski) [inline extensions/wikia/Flags/Flags.setup.php]: Okey dokey! :+1:

File: extensions/wikia/Flags/Flags.setup.php (hunk 1 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/Flags.setup.php]: :+1:

File: extensions/wikia/Flags/models/Flag.class.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		/**
+		 * The data should be verified in the
+		 * verifyParamsForAction() method before
+		 * performing an action
+		 */
+		if ( !$this->paramsVerified ) {
+			return false;
+		}
+
+		switch ( $action ) {
```
Comment (lgarczewski) [inline extensions/wikia/Flags/models/Flag.class.php]: Not sure we want to use a switch here, using different methods and possibly building them based on the value of the action variable may be an alternative solution.

File: extensions/wikia/Flags/controllers/FlagsController.class.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+class FlagsController extends WikiaController {
+
+	private
+		$action,
+		$model,
+		$params,
+		$status = false;
+
+	public function getFlagsForPage() {
+		$this->skipRendering();
```
Comment (Grunny) [inline extensions/wikia/Flags/controllers/FlagsController.class.php]: Why is this being called on each of these methods?

File: extensions/wikia/Flags/models/Flag.class.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+				$this->status = $this->removeFlag( $params['flagId'] );
+				break;
+		}
+
+		return $this->status;
+	}
+
+	public function verifyParamsForAction( $action, $params ) {
+		$this->debug( [ $action, $params ] );
+		switch ( $action ) {
```
Comment (lgarczewski) [inline extensions/wikia/Flags/models/Flag.class.php]: Same as above -- possibly combine verify and perform and make it per action type? This seems like complicating the code if I'm reading it and trying to follow.

File: extensions/wikia/Flags/controllers/FlagsController.class.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		$this->model = new Flag();
+
+		return $this->model->getFlagsForPage( $this->params['wikiId'], $this->params['pageId'] );
+	}
+
+	public function getFlag() {
+
+	}
+
+	public function flagAction() {
```
Comment (Grunny) [inline extensions/wikia/Flags/controllers/FlagsController.class.php]: This method is performing a write action so must require a POST and verify a user token to protect against CSRF.

File: extensions/wikia/Flags/controllers/FlagsController.class.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+		if ( $this->model->verifyParamsForAction( $this->action, $this->params ) ) {
+			$this->status = $this->model->performAction( $this->action, $this->params );
+		}
+
+		$this->setVal( 'status', $this->status );
+		return $this->status;
+	}
+
+	public function flagTypeAction() {
```
Comment (Grunny) [inline extensions/wikia/Flags/controllers/FlagsController.class.php]: Likewise, this method is performing a write action so must require a POST and verify a user token to protect against CSRF.

File: extensions/wikia/Flags/models/Flag.class.php [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		}
+
+		if ( !empty( $flagsWithParams ) ) {
+			$flagsParams = $this->getParamsForFlags( $flagsWithParams );
+			foreach ( $flagsParams as $flagId => $flagParams ) {
+				$flagsForPage[ $flagId ]['params'] = $flagParams;
+			}
+		}
+
+		var_dump( $flagsForPage );
```
Comment (Grunny) [inline extensions/wikia/Flags/models/Flag.class.php]: Testing code?

File: extensions/wikia/Flags/models/Flag.class.php [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+			->INSERT( self::FLAGS_TO_PAGES_TABLE )
+				// flag_id is auto_increment
+				->SET( 'flag_type_id', $flagTypeId )
+				->SET( 'wiki_id', $wikiId )
+				->SET( 'page_id', $pageId )
+			->run( $db );
+
+		$db->commit();
+
+		$status = $db->affectedRows() > 0;
```
Comment (Grunny) [inline extensions/wikia/Flags/models/Flag.class.php]: `$db->affectedRows()` will always be 0 here as it is being called after `$db->commit();`. You'll need to get the value from before the commit.

File: extensions/wikia/Flags/models/Flag.class.php [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+		$db = $this->getDatabaseForWrite();
+
+		$sql = ( new \WikiaSQL() )
+			->DELETE( self::FLAGS_TO_PAGES_TABLE )
+			->WHERE( 'flag_id' )->EQUAL_TO( $flagId )
+			->run( $db );
+
+		$db->commit();
+
+		$status = $db->affectedRows() > 0;
```
Comment (Grunny) [inline extensions/wikia/Flags/models/Flag.class.php]: Same here, `$db->affectedRows()` will always be 0 as it is being called after `$db->commit();`. You'll need to get the value from before the commit.

File: extensions/wikia/Flags/models/FlagParameter.class.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+		foreach ( $params as $paramName => $paramValue ) {
+			$sql->VALUES( $flagId, $wikiId, $pageId, $flagTypeId, $paramName, $paramValue );
+		}
+
+		$sql->run( $db );
+
+		$db->commit();
+
+		$this->status = $db->affectedRows() > 0;
```
Comment (Grunny) [inline extensions/wikia/Flags/models/FlagParameter.class.php]: And here.

File: extensions/wikia/Flags/models/FlagType.class.php [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+			->SET( 'flag_view', $flagView )
+			->SET( 'flag_targeting', $flagTargeting );
+			if ( $flagParamsNames !== null  ) {
+				$sql->SET('flag_params_names', $flagParamsNames);
+			}
+			$sql->run( $db );
+
+		$db->commit();
+
+		$this->status = $db->affectedRows() > 0;
```
Comment (Grunny) [inline extensions/wikia/Flags/models/FlagType.class.php]: And here.

File: extensions/wikia/Flags/models/FlagType.class.php [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+		$db = $this->getDatabaseForWrite();
+
+		$sql = ( new \WikiaSQL() )
+			->DELETE( self::FLAGS_TO_PAGES_TABLE )
+			->WHERE( 'flag_type_id' )->EQUAL_TO( $flagType )
+			->run( $db );
+
+		$db->commit();
+
+		$this->status = $db->affectedRows() > 0;
```
Comment (Grunny) [inline extensions/wikia/Flags/models/FlagType.class.php]: And finally here. ;)

File: extensions/wikia/Flags/models/FlagsModel.class.php [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Flags\Models;
+
+class FlagsModel extends \WikiaModel {
+
+	const FLAGS_TO_PAGES_TABLE = 'flags_to_pages';
+	const FLAGS_TYPES_TABLE = 'flags_types';
+	const FLAGS_PARAMS_TABLE = 'flags_params';
+
+	protected function getDatabaseForRead() {
```
Comment (Grunny) [inline extensions/wikia/Flags/models/FlagsModel.class.php]: I'm not sure of the value of these over the method WikiaModel already has:
`$this->getSharedDB();`
`$this->getSharedDB( DB_MASTER );`

File: extensions/wikia/Flags/models/Flag.class.php (hunk 8 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/models/Flag.class.php]: Yup :)

File: extensions/wikia/Flags/models/FlagType.class.php (hunk 13 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/models/FlagType.class.php]: Oh, you're stubborn (thank God!) :)

File: extensions/wikia/Flags/models/FlagsModel.class.php (hunk 14 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/models/FlagsModel.class.php]: The SharedDB is a temporary solution until we get our own DB. Then the methods will be Flags specific.

File: extensions/wikia/Flags/controllers/FlagsController.class.php (hunk 7 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/controllers/FlagsController.class.php]: :+1:

File: extensions/wikia/Flags/models/Flag.class.php (hunk 5 above)
Comment (adamkarminski) [inline extensions/wikia/Flags/models/Flag.class.php]: Got it boss!

File: extensions/wikia/Flags/models/FlagsModel.class.php (hunk 14 above)
Comment (Grunny) [inline extensions/wikia/Flags/models/FlagsModel.class.php]: :+1:

File: extensions/wikia/Flags/models/FlagType.class.php (hunk 13 above)
Comment (lgarczewski) [inline extensions/wikia/Flags/models/FlagType.class.php]: Well spotted, sir.

File: extensions/wikia/Flags/models/FlagType.class.php [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+namespace Flags\Models;
+
+class FlagType extends FlagsModel {
+	const FLAG_TYPE_ACTION_ADD = 'add';
+	const FLAG_TYPE_ACTION_REMOVE = 'remove';
+
+	private
+		$status,
+		$paramsVerified = false,
+		$flagGroups = [
```
Comment (lukaszkonieczny) [inline extensions/wikia/Flags/models/FlagType.class.php]: What do you think about making them public?

File: extensions/wikia/Flags/models/Flag.class.php [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+		if ( $status && !empty( $params ) ) {
+			$flagId = $db->insertId();
+			$paramsModel = new FlagParameter();
+			$status = $paramsModel->createParametersForFlag( $flagId, $wikiId, $pageId, $flagTypeId, $params );
+		}
+
+		return $status;
+	}
+
+	private function removeFlag( $flagId, $removeParams = false ) {
```
Comment (lukaszkonieczny) [inline extensions/wikia/Flags/models/Flag.class.php]: I think we should remove params when the flag is removed

File: extensions/wikia/Flags/models/FlagType.class.php (hunk 15 above)
Comment (lukaszkonieczny) [inline extensions/wikia/Flags/models/FlagType.class.php]: But...I'm thinking now, how to easily get flag name by id?

File: extensions/wikia/Flags/models/FlagType.class.php [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+		7 => 'Article status',
+		8 => 'For readers',
+		9 => 'For contributors',
+	];
+
+	public function getFlagGroupsMapping() {
+		return self::$flagGroups;
+	}
+
+	public function getFlagTypesForWikia( $wikiId ) {
```
Comment (adamkarminski) [inline extensions/wikia/Flags/models/FlagType.class.php]: A placeholder method.

**Comment by adamkarminski** [comment]
everything's cool :)
~Tor

**Pull request merged**